src/compiler/transformers/declarations.ts

```diff
@@ -1657,38 +1657,34 @@ export function transformDeclarations(context: TransformationContext): Transform
 
                 const extendsClause = getEffectiveBaseTypeNode(input);
                 if (extendsClause && !isEntityNameExpression(extendsClause.expression) && extendsClause.expression.kind !== SyntaxKind.NullKeyword) {
-                    // We must add a temporary declaration for the extends clause expression
+                    // For classes which get their base type from an expression other than a name or Null, replace the expression they extend with one of the form:
+                    // declare class A extends ({} as DEFINITION_OF_BASE_HERE) {}
 
-                    const oldId = input.name ? unescapeLeadingUnderscores(input.name.escapedText) : "default";
-                    const newId = factory.createUniqueName(`${oldId}_base`, GeneratedIdentifierFlags.Optimistic);
                     getSymbolAccessibilityDiagnostic = () => ({
                         diagnosticMessage: Diagnostics.extends_clause_of_exported_class_0_has_or_is_using_private_name_1,
                         errorNode: extendsClause,
                         typeName: input.name,
                     });
-                    const varDecl = factory.createVariableDeclaration(newId, /*exclamationToken*/ undefined, resolver.createTypeOfExpression(extendsClause.expression, input, declarationEmitNodeBuilderFlags, declarationEmitInternalNodeBuilderFlags, symbolTracker), /*initializer*/ undefined);
-                    const statement = factory.createVariableStatement(needsDeclare ? [factory.createModifier(SyntaxKind.DeclareKeyword)] : [], factory.createVariableDeclarationList([varDecl], NodeFlags.Const));
+                    const baseType = resolver.createTypeOfExpression(extendsClause.expression, input, declarationEmitNodeBuilderFlags, declarationEmitInternalNodeBuilderFlags, symbolTracker);
+                    const extendsExpression = factory.createAsExpression(factory.createObjectLiteralExpression(), baseType!); // TODO: GH#18217

```

Comment:
Having never worked on this code before, and the relevant types and methods mostly being undocumented, I don't know why this `createTypeOfExpression` might return undefined or if ignoring undefined values here is safe to do here.

I'm also not sure if putting the `// TODO: GH#18217` makes sense or not: it seemed to be used to track questionable use of `!` (like the one I removed below) so I left it here to help draw attention to the `!` here.